server/src/main/java/org/opensearch/indices/replication/checkpoint/PublishCheckpointAction.java

```diff
@@ -8,14 +8,166 @@
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.index.IndexNotFoundException;
+import org.opensearch.index.shard.IndexShard;
+import org.opensearch.index.shard.IndexShardClosedException;
+import org.opensearch.indices.IndicesService;
+import org.opensearch.indices.replication.SegmentReplicationReplicaService;
+import org.opensearch.indices.replication.copy.PrimaryShardReplicationSource;
+import org.opensearch.node.NodeClosedException;
+import org.opensearch.tasks.Task;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportException;
+import org.opensearch.transport.TransportResponseHandler;
+import org.opensearch.transport.TransportService;
 
-public class PublishCheckpointAction extends ActionType<RefreshResponse> {
-    public static final PublishCheckpointAction INSTANCE = new PublishCheckpointAction();
-    public static final String NAME = "indices:admin/publishCheckpoint";
+import java.io.IOException;
+import java.util.Objects;
 
-    private PublishCheckpointAction() {
-        super(NAME, RefreshResponse::new);
+public class PublishCheckpointAction extends TransportReplicationAction<
+    PublishCheckpointRequest,
+    PublishCheckpointRequest,
+    ReplicationResponse> {
+
+    public static final String ACTION_NAME = "indices:admin/publishCheckpoint";
+    protected static Logger logger = LogManager.getLogger(PublishCheckpointAction.class);
+
+    private final SegmentReplicationReplicaService replicationService;
+    private final PrimaryShardReplicationSource source;
+
+    @Inject
+    public PublishCheckpointAction(
+        Settings settings,
+        TransportService transportService,
+        ClusterService clusterService,
+        IndicesService indicesService,
+        ThreadPool threadPool,
+        ShardStateAction shardStateAction,
+        ActionFilters actionFilters,
+        SegmentReplicationReplicaService segmentCopyService,
+        PrimaryShardReplicationSource source
+    ) {
+        super(
+            settings,
+            ACTION_NAME,
+            transportService,
+            clusterService,
+            indicesService,
+            threadPool,
+            shardStateAction,
+            actionFilters,
+            PublishCheckpointRequest::new,
+            PublishCheckpointRequest::new,
+            ThreadPool.Names.REFRESH
+        );
+        this.replicationService = segmentCopyService;
+        this.source = source;
+    }
+
+    @Override
+    protected ReplicationResponse newResponseInstance(StreamInput in) throws IOException {
+        return new ReplicationResponse(in);
+    }
+
+    @Override
+    protected void doExecute(Task task, PublishCheckpointRequest request, ActionListener<ReplicationResponse> listener) {
+        assert false : "use PublishCheckpointAction#publish";
```

Comment:
What's the purpose of this function if it throws an assertion error everytime?

Reply:
Yeah, the implementation here isn't the clearest. What's going on is:

1. To allow a primary shard to publish the checkpoint to its replicas, we're leveraging the ReplicationOperation functionality of executing on a primary followed by its replicas.
2. In order to run a ReplicationOperation, we need to register an action name with the transport layer. TransportReplicationAction (the superclass here) does this in its constructor
3. _However_, TransportReplicationAction also assumes that it will be called via execute/doExecute, which includes rerouting behavior. We don't need the reroute phase here since we are guaranteed to be on the primary shard
4. Thus, the purpose of this implementation is to guard against accidental execution via the above code path. Instead, the refresh listener invokes the _publish_ API directly